### PR TITLE
Add hook reference generator README and documentation

### PR DESCRIPTION
**Pull request by psealock**
Add a README to the hook reference generator. While the script and resulting data aren't currently used, I'd like to use this PR to solicit any feedback on the mechanism and data before appropriately documenting all JS filters in the repo.

I also removed the `example` tag because it doesn't offer enough useful information to justify contriving an example use case for every filter. Instead, I gather all `param` tags which will be more useful.

### Detailed test instructions:

Make sure the README makes sense, try out the script, and see the JSON object satisfies needs for documentation.

### CODE REVIEW

**Review by chihsuan — COMMENTED** [review]
@psealock Thanks for working on this. LGTM and test well. However, I'm not familiar with the hook reference generator so I'll leave others to approve this.

**Review by ilyasfoo — COMMENTED** [review]
Overall this looks good! Some comments:

File: packages/onboarding/src/components/WooOnboardingTaskListItem/WooOnboardingTaskListItem.js [hunk 1]
```diff
@@ -4,6 +4,14 @@
 import { createElement } from '@wordpress/element';
 import { Slot, Fill } from '@wordpress/components';
 
+/**
+ * A Fill for adding Onboarding Task List items.
+ *
+ * @slotFill WooOnboardingTaskListItem
+ * @scope wwoocommerce-tasks
```
[the remark this answers is not in the document]
Comment (louwie17) [reply packages/onboarding/src/components/WooOnboardingTaskListItem/WooOnboardingTaskListItem.js]: It appears this wasn't intended as the scope contains one `w` -> https://github.com/woocommerce/woocommerce-admin/blob/main/client/layout/index.js#L181

File: bin/hook-reference/data.js [hunk 2]
```diff
@@ -53,16 +53,36 @@ const prepareHooks = async ( path ) => {
...
 			sourceFile,
 			name: tag ? tag.name : '',
-			example: example ? example.description : '',
 			type: tag.tag,
+			params: paramTags,
 		};
+
+		if ( tag.tag === 'slotFill' ) {
+			const scopeTab = tags.find( ( tag ) => tag.tag === 'scope' );
+			docObject.scope = scopeTab.name;
```
[the remark this answers is not in the document]
Comment (louwie17) [reply bin/hook-reference/data.js]: I fixed this by making it optional, but outputting a warning.

File: client/analytics/report/categories/config.js [hunk 3]
```diff
@@ -118,6 +143,23 @@ if ( Object.keys( advancedFilters.filters ).length ) {
...
+/**
+ * An object defining a set of report filters.
+ *
+ * @typedef {Object} filter
+ * @property {string} label Label describing the set of filters.
+ * @property {string} param Url query param this set of filters operates on.
+ * @property {Array.<string>} staticParams Array of `param` that remain constant when other params are manipulated.
+ * @property {Function} showFilters A function with url query as an argument returning a Boolean depending on whether or not the filters should be shown.
+ * @property {Array} filters An array of filter objects.
+ */
```
[the remark this answers is not in the document]
Comment (louwie17) [reply client/analytics/report/categories/config.js:155]: I updated it to use the @import method as you suggested, we do seem to be using this in a couple other spots.

**Comment by louwie17** [comment]
@ilyasfoo this is ready for another review whenever you have a change, I took over and updated it as I noticed that Paul is on sabbatical, and it will be nice to have this merged prior to that.

**Review by ilyasfoo — APPROVED** [review]
@louwie17 Thanks so much for addressing this while Paul is away!

One super nitpicky comment is I feel slightly odd with the typedef imports from various reports referring to `categories` report instead of living in a "shared" file.  But I digress, ignore this comment at will! Perhaps typescript would've been a better solution anyway.

I've re-tested it, works well and looking great!  🚢

**Comment by louwie17** [comment]
> One super nitpicky comment is I feel slightly odd with the typedef imports from various reports referring to `categories` report instead of living in a "shared" file. But I digress, ignore this comment at will! Perhaps typescript would've been a better solution anyway.

@ilyasfoo I had a similar thought, although it also felled weird to me to create a separate file just for JSDoc definitions. And I didn't see a better file to move the definitions into otherwise (although I didn't look that hard).
Looking at it quickly again, I could move them to the `report/index.js` file to keep it in a more hierarchical place.

**Comment by ilyasfoo** [comment]
>Looking at it quickly again, I could move them to the report/index.js file to keep it in a more hierarchical place.

@louwie17 Sounds good if you'd like to make the changes! Otherwise, I'm good too

**Comment by louwie17** [comment]
> @louwie17 Sounds good if you'd like to make the changes! Otherwise, I'm good too

@ilyasfoo I meant to do this yesterday, but didn't get to it. I just pushed the changes up if you want to give it another quick glance?

File: client/analytics/report/index.js [hunk 4]
```diff
@@ -21,6 +21,28 @@ import {
 } from '../../lib/currency-context';
 import getReports from './get-reports';
 
+/**
+ * An object defining a report page.
```
Comment (ilyasfoo) [inline client/analytics/report/index.js:25]:
```suggestion
 * An object defining a chart.
```

File: client/store-management-links/index.js [hunk 5]
```diff
@@ -159,6 +159,21 @@ export const StoreManagementLinks = () => {
 	const shopUrl = getAdminSetting( 'shopUrl' );
 
 	const extensionQuickLinks = generateExtensionLinks(
+		/**
+		 * An object defining a report page.
```
Comment (ilyasfoo) [inline client/store-management-links/index.js:163]:
```suggestion
		 * An object defining an extension link.
```

File: client/tasks/fills/products/product-template-modal.js [hunk 6]
```diff
@@ -133,6 +133,21 @@ export default function ProductTemplateModal( { onClose } ) {
 		  )
 		: getProductTemplates();
 
+	/**
+	 * An object defining a report page.
```
Comment (ilyasfoo) [inline client/tasks/fills/products/product-template-modal.js:137]:
```suggestion
	 * An object defining a product template.
```

**Review by ilyasfoo — APPROVED** [review]
Thanks for making the change, @louwie17!

I've found another instance that's also duplicate not re-using the typedef:

https://github.com/woocommerce/woocommerce-admin/blob/46c8304c425749dfc715b38e59f56198b05e7b46/client/dashboard/customizable.js#L36-L45

Also, unrelated to your fixes, I've found a couple more typedefs with inaccurate comments. I've put my suggestions for them.

Pre-approving! 🚢

**Pull request merged**